### DISCUSSION TITLE
X39 thermostat

### DISCUSSION THREAD

**Discussion by gefkuz** [arendst/Tasmota#14550] [Ideas]
Hello, I'm using since few weeks the thermostat draiver of tasmota with an esp8266 and a thermostatic valve. I like working with PIDs. But with this feature, the minimim tick time is minutes. Is that possible to make modifications to work in seconds and then have a more precisely heating T° on the heaters and have always a water flux in all heaters? Tx

**Comment by sfromis** [comment]
Not using PID, but in my understanding the update period is TelePeriod which you can select, down to 10 seconds.

**Reply by barbudor** [reply]
@gefkuz are you refering to `TimePiCycleSet` command which unit is in minutes ?

Fomr the code, I see it is limited between 0 (?) and 1440 minutes and stored in a uint16_t
However, this is not stored in flash settings so it should be possible to introduce a non-breaking change.
I can have a look later tonight

I have no idea what would be the impact on the rest of the code. and what should be the minimum to enfore.
How would the thermostat behave with value down to 30 seconds ? 10 seconds ? below 10 seconds ?

**Reply by gefkuz** [reply]
Yes i'm refering to TimePiCycleSet. But not only.
I'm seeing that state change of the output is multiple of minutes. never 1m10s but always 1m or 2m or xminutes.
if all times variables can be in seconds that's wonderfull...

**Reply by gefkuz** [reply]
All thermostats are different I guess. The real question is: the tasmota host ic like esp8266 or esp32 is capable to drive a PWM down to 10 seconds? I don't how is made the code of tasmota. 
My situation, the thermostat is not fast. When activated, it take 5 minutes to open the valve and 2 minutes to close it when deactivated.
A heating system don't like valve closed completely because there is pump in the system. And if all valves are closed, the pump is in trouble and wear  out.
If we make a signal on and off with a period minus than time on and time off, then we can control the opening valve pourcentage.
And then the pump of the system is less in trouble.
Please excuse my english. And tell me if you want more explanation.

**Reply by techman83** [reply]
I do have a use case for higher frequency PID. It's a common thing to add a PID to Coffee machines like the Gaggia classic. Very likely you'll overshoot/undershoot the ideal brew temperature, being that the ideal brew time is 20-30 seconds.

Admittedly it wouldn't be out of the question to add a dedicated device with PID loaded up, then leave the control functions to Tasmota, but it would be neat to have it all in one package.

**Reply by gefkuz** [reply]
@techman83 , the driver x39 thermostat already have a PID function.
step time is 1 minute. That's the point..... If the the step time can be 1 second, we can control more precisely temperature and limit overshhot/undershoot. And have a more stable temperature.

**Reply by barbudor** [reply]
@gefkuz 
Can you please provide me your user_config_override.h so I can build a test version for you ?
Or will you be able to build a test bin yourself from my fork ?

**Reply by gefkuz** [reply]
I'm not able to compile tasmota with tour fork.
What's this file config? I'm using an esp8266 wemos witwh thermostat that's all.

**Reply by barbudor** [reply]
@gefkuz Which binary are you using exactly ? from where 

It's a little more complicated than I thought because I'm not sure it makes sense to change TimePiCycle without also changing TimeMaxAction and TimeMinAction


Hi @arijav would you be able to advise or help ? @gefkuz would like to have shorter cycle time than 1 minute so I was in the process to see how to change `int16_t time_pi_cycle` into `int32_t time_pi_cycle_secs` but I have 2 questions:
1. Should also `time_max_action` and `time_min_action` be upgraded to seconds ?
2. I'm confused by the line: https://github.com/arendst/Tasmota/blob/10e480bfd9cb80057278442c1c6074614bc38673/tasmota/xdrv_39_thermostat.ino#L1125
`TasmotaGlobal.uptime` is in seconds.
But you are adding `time_pi_cycle * dutycycle_step_autotune / 100` where 
`time_pi_cycle` is in minutes
`dutycycle_step_autotune` is between 0 and 100 (uint8_t)
The units doesn't seems to align correctly
So I was wondering if I should just replace `time_pi_cycle` by my new `time_pi_cycle_secs` or if I should add a further divide by 60 ?

Thanks

**Reply by barbudor** [reply]
@gefkuz Please try to upgrade with the attached test version.
It should be equivalent to the tasmota-thermostat binary from Jason's unofficial build
Command `TimePiCycle` can now accept a parameter in seconds if you add `sec` in the command line such as:
`TimePiCycle 30 sec` (there could be a space or not between the number and `sec`)
On the web UI the display has been adapted to:
![image](https://user-images.githubusercontent.com/2996042/150643652-6d3ff691-2b82-4144-87ad-f41dc3d6fcf7.png)

[tasmota-thermostat-test.zip](https://github.com/arendst/Tasmota/files/7918598/tasmota-thermostat-test.zip)

I don't have enough experience with the Thermostat feature yet to be sure that this is 100% safe, so please test in deep.
It will not be pushed to Tasmota before someone with experience can review the change.

Hi @arijav 
You feedback would be more than welcomed
The changes are in this branch https://github.com/barbudor/Tasmota/tree/therm_pi_cycle_in_secs

**Reply by gefkuz** [reply]
Hello, wwhen I see this function : bool ThermostatMinuteCounter(uint8_t ctr_output)
In this function, we made the system working on minutes.
It seems that if this function doesn't block the system to work in minute but at each second, it works for all parameters.

**Reply by barbudor** [reply]
Looks like I under estimated the changes to do.
Indeed I don't have a good understanding on how the module is working.
It would be preferable to let the original author to advise on this.

**Reply by gefkuz** [reply]
ok @barbudor . thanks for you help ;)
